diff --git a/arch/arm/mach-bcm283x/Kconfig b/arch/arm/mach-bcm283x/Kconfig
index d9303e8c281..e7e47b1166d 100644
--- a/arch/arm/mach-bcm283x/Kconfig
+++ b/arch/arm/mach-bcm283x/Kconfig
@@ -56,6 +56,10 @@ config BCM2711_64B
 	select BLOBLIST_ALLOC if GENERATE_ACPI_TABLE
 	select BLOBLIST_TABLES if GENERATE_ACPI_TABLE
 
+config BCM2712
+	bool "Broadcom BCM2712 SoC support"
+	depends on ARCH_BCM283X
+
 menu "Broadcom BCM283X family"
 	depends on ARCH_BCM283X
 choice

diff --git a/arch/arm/mach-bcm283x/init.c b/arch/arm/mach-bcm283x/init.c
index 7a1de22e0ae..e72bd6af191 100644
--- a/arch/arm/mach-bcm283x/init.c
+++ b/arch/arm/mach-bcm283x/init.c
@@ -18,7 +18,7 @@
 #ifdef CONFIG_ARM64
 #include <asm/armv8/mmu.h>
 
-#define MEM_MAP_MAX_ENTRIES (4)
+#define MEM_MAP_MAX_ENTRIES (6)
 
 static struct mm_region bcm283x_mem_map[MEM_MAP_MAX_ENTRIES] = {
 	{
@@ -83,6 +83,22 @@ static struct mm_region bcm2712_mem_map[MEM_MAP_MAX_ENTRIES] = {
 		.attrs = PTE_BLOCK_MEMTYPE(MT_DEVICE_NGNRNE) |
 			 PTE_BLOCK_NON_SHARE |
 			 PTE_BLOCK_PXN | PTE_BLOCK_UXN
+	}, {
+		/* Beginning of PCIe section for NVMe */
+		.virt = 0x1b80000000UL,
+		.phys = 0x1b80000000UL,
+		.size = 0x80000000UL,
+		.attrs = PTE_BLOCK_MEMTYPE(MT_DEVICE_NGNRNE) |
+			 PTE_BLOCK_NON_SHARE |
+			 PTE_BLOCK_PXN | PTE_BLOCK_UXN
+	}, {
+		/* Beginning of PCIe section */
+		.virt = 0x1f00000000UL,
+		.phys = 0x1f00000000UL,
+		.size = 0x0002000000UL,
+		.attrs = PTE_BLOCK_MEMTYPE(MT_DEVICE_NGNRNE) |
+			 PTE_BLOCK_NON_SHARE |
+			 PTE_BLOCK_PXN | PTE_BLOCK_UXN
 	}, {
 		/* SoC bus */
 		.virt = 0x107c000000UL,

diff --git a/board/raspberrypi/rpi/rpi.c b/board/raspberrypi/rpi/rpi.c
index f9b643555dd..6b12361b526 100644
--- a/board/raspberrypi/rpi/rpi.c
+++ b/board/raspberrypi/rpi/rpi.c
@@ -20,6 +20,7 @@
 #include <asm/arch/sdhci.h>
 #include <asm/global_data.h>
 #include <dm/platform_data/serial_bcm283x_mu.h>
+#include <broadcom/bcm_board_types.h>
 #ifdef CONFIG_ARM64
 #include <asm/armv8/mmu.h>
 #endif
@@ -532,6 +533,10 @@ static void get_board_revision(void)
 		model = &models[rev_type];
 	}
 
+#ifdef CONFIG_BOARD_TYPES
+	gd->board_type = rev_type;
+#endif
+
 	printf("RPI %s (0x%x)\n", model->name, revision);
 }
 
@@ -644,6 +649,29 @@ int ft_board_setup(void *blob, struct bd_info *bd)
 	return 0;
 }
 
+/* TODO: Using late_init to initialize pci device with ID_RP1.
+ * RP1 pci device should be initialized by the PCI subsystem because
+ * it is under develop right now and depends from the final device-tree
+ * format from the Linux Kernel. Current device-tree format violates
+ * pci driver model. So this should be changed after upstreaming RP1
+ * to the Linux Kernel source code.
+ * This initialization should be done only for RPI5 board.
+ */
+#ifdef CONFIG_BCM2712
+int board_late_init(void)
+{
+	struct udevice *dev;
+	int err;
+
+	/* Only initialize RP1 on Raspberry Pi 5 */
+	if (gd_board_type() != BCM2712_RPI_5_B_NEW)
+		return 0;
+
+	err = dm_pci_find_device(PCI_VENDOR_ID_RPI, PCI_DEVICE_ID_RP1_C0,
+				 0, &dev);
+	if (err)
+		printf("RPI: RP1 device not found\n");
+
+	return 0;
+}
+#endif

 #if CONFIG_IS_ENABLED(GENERATE_ACPI_TABLE)
 static bool is_rpi4(void)
 {

diff --git a/board/raspberrypi/rpi/rpi.env b/board/raspberrypi/rpi/rpi.env
index 9ac9d6768ca..f879df109ac 100644
--- a/board/raspberrypi/rpi/rpi.env
+++ b/board/raspberrypi/rpi/rpi.env
@@ -77,4 +77,4 @@ pxefile_addr_r=0x05500000
 fdt_addr_r=0x05600000
 ramdisk_addr_r=0x05700000
 
-boot_targets=mmc usb pxe dhcp
+boot_targets=mmc nvme usb pxe dhcp

diff --git a/include/broadcom/bcm_board_types.h b/include/broadcom/bcm_board_types.h
new file mode 100644
index 00000000000..69db528874e
--- /dev/null
+++ b/include/broadcom/bcm_board_types.h
@@ -0,0 +1,25 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * This header defines BCM board types.
+ *
+ * Copyright (c) 2024 EPAM Systems
+ *
+ */
+#ifndef _BCM_BOARD_TYPES_H_
+#define _BCM_BOARD_TYPES_H_
+
+/*
+ * TODO: This enumerator defines the main BCM types.
+ * These types are defined by the BCM firmware and requested
+ * via mbox. Currently only one board type was added.
+ * This is needed to determine whether current board is bcm2712
+ * or other to control the behaviour of the following drivers:
+ * macb and pcie_brcmstb. New board enumerators can be added here
+ * if needed.
+ * See https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#raspberry-pi-revision-codes
+ */
+typedef enum {
+	BCM2712_RPI_5_B_NEW = 0x17,
+} bcm_board_type_t;
+
+#endif /* _BCM_BOARD_TYPES_H_ */
